### PR TITLE
tests: lib: cmsis:transform: cf64 min_ram 128k

### LINKED ISSUE #27849: tests: lib: cmsis_dsp: transform: malloc out of memory
**Describe the bug**
When running tests/lib/cmsis_dsp/transform/ on nucleo_f207zg,
with config: libraries.cmsis_dsp.transform.cf64 ,
`malloc()` failed as it runs out of memory.
Please note that nucleo_f207zg has 128k RAM (whereas there is a min_ram of 64)

**To Reproduce**
1. sanitycheck -N --device-testing --hardware-map ../map.yaml -c -T tests/lib/cmsis_dsp/transform/ -p nucleo_f207zg
2. See error

**Expected behavior**
Test passed

**Logs and console output**
```
*** Booting Zephyr OS build zephyr-v2.3.0-1075-g4e12dfb5ebda  ***
Running test suite transform_cf64
===================================================================
START - test_arm_cfft_f64_cmplx_noisy_16
 PASS - test_arm_cfft_f64_cmplx_noisy_16
===================================================================
START - test_arm_cifft_f64_cmplx_noisy_16
 PASS - test_arm_cifft_f64_cmplx_noisy_16
===================================================================
START - test_arm_cfft_f64_cmplx_noisy_32
 PASS - test_arm_cfft_f64_cmplx_noisy_32
===================================================================
START - test_arm_cifft_f64_cmplx_noisy_32
 PASS - test_arm_cifft_f64_cmplx_noisy_32
===================================================================
START - test_arm_cfft_f64_cmplx_noisy_64
 PASS - test_arm_cfft_f64_cmplx_noisy_64
===================================================================
START - test_arm_cifft_f64_cmplx_noisy_64
 PASS - test_arm_cifft_f64_cmplx_noisy_64
===================================================================
START - test_arm_cfft_f64_cmplx_noisy_128
 PASS - test_arm_cfft_f64_cmplx_noisy_128
===================================================================
START - test_arm_cifft_f64_cmplx_noisy_128
 PASS - test_arm_cifft_f64_cmplx_noisy_128
===================================================================
START - test_arm_cfft_f64_cmplx_noisy_256
 PASS - test_arm_cfft_f64_cmplx_noisy_256
===================================================================
START - test_arm_cifft_f64_cmplx_noisy_256
 PASS - test_arm_cifft_f64_cmplx_noisy_256
===================================================================
START - test_arm_cfft_f64_cmplx_noisy_512
 PASS - test_arm_cfft_f64_cmplx_noisy_512
===================================================================
START - test_arm_cifft_f64_cmplx_noisy_512
 PASS - test_arm_cifft_f64_cmplx_noisy_512
===================================================================
START - test_arm_cfft_f64_cmplx_noisy_1024
 PASS - test_arm_cfft_f64_cmplx_noisy_1024
===================================================================
START - test_arm_cifft_f64_cmplx_noisy_1024
 PASS - test_arm_cifft_f64_cmplx_noisy_1024
===================================================================
START - test_arm_cfft_f64_cmplx_noisy_2048
 PASS - test_arm_cfft_f64_cmplx_noisy_2048
===================================================================
START - test_arm_cifft_f64_cmplx_noisy_2048
 PASS - test_arm_cifft_f64_cmplx_noisy_2048
===================================================================
START - test_arm_cfft_f64_cmplx_noisy_2048
 PASS - test_arm_cfft_f64_cmplx_noisy_2048
===================================================================
START - test_arm_cifft_f64_cmplx_noisy_2048
 PASS - test_arm_cifft_f64_cmplx_noisy_2048
===================================================================
START - test_arm_cfft_f64_cmplx_noisy_4096
Assertion failed at WEST_TOPDIR/zephyr/tests/lib/cmsis_dsp/transform/src/cf64.c:27: test_arm_cfft_f64_cmplx_backend: (output is NULL)
buffer allocation failed
 FAIL - test_arm_cfft_f64_cmplx_noisy_4096
```

**Environment (please complete the following information):**
 - OS: Linux and Windows)
 - Toolchain Zephyr SDK
 - Commit SHA: 70ff1d73c42f2e174c49b1e217712b52ad6ae78b

**Analysis**
`malloc()` fails as it runs out of memory when it reaches step: `test_arm_cfft_f64_cmplx_noisy_4096`

&nbsp;&nbsp;&nbsp;&nbsp;`DEFINE_TEST_VARIANT4(arm_cifft_f64_cmplx, step_4096, &arm_cfft_sR_f64_len4096,`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`	in_cifft_step_4096, in_cfft_step_4096,  `**8192**);

Analysing further I faced 2 issues:

1) testcase.yaml has a min ram property for this config
`    min_ram: 64`
But while executing step `test_arm_cfft_f64_cmplx_noisy_4096` there is a malloc() of length:
8192 * sizeof(float64_t) = 8192 * 8 = 64k
but we need more RAM for variables, stack, ...
so 64K RAM is no enough.

2) I don't know exactly how malloc() algorithm from newlibc is working,
but it appears during my debug session that even if memory is freed between each step,
each time the size of the malloc is increased, the previous memory size is 'lost' (probably due to heap fragmentation).
So with current implementation before reaching step 4096, we loose 256 + 512 + 1024 + 2048 + ... + 16k + 32k, which is almost 64k lost. And  then it is no more possible to request 64k besides. Finally my 128k RAM are not enough.
If I reverse the order of the test steps : starting from 4096 and decreasing to 16, then test is passed.
Probably malloc() algorithm is able to reuse big memory chunk for smaller malloc which reduce fragmentation.

So 128k is  sufficient to run the test, but it needs changes.

Note: there is no problem with config libraries.cmsis_dsp.transform.rf64 because the max malloc is for `test_arm_rifft_f64_real_noisy_4096`
4096 * sizeof(float64_t) = 4096 * 8 = 32k

### PR DESCRIPTION
**Pull request by ABOSTM**
tests: lib: cmsis:transform: cf64 min_ram 128k

tests/lib/cmsis_dsp/transform/ with config:
   * libraries.cmsis_dsp.transform.cf64
Needs more than 64k RAM.
128k is sufficient if tests steps are reordered
(decreasing malloc size)

Fixes #27849

### CODE REVIEW

File: tests/lib/cmsis_dsp/transform/testcase.yaml [hunk 1]
```diff
@@ -47,7 +47,7 @@ tests:
       - CONFIG_CMSIS_DSP_TEST_TRANSFORM_RF32=y
   libraries.cmsis_dsp.transform.cf64:
     min_flash: 1024
-    min_ram: 64
+    min_ram: 128
```
Comment (stephanosio) [inline tests/lib/cmsis_dsp/transform/testcase.yaml:50]: > But while executing step test_arm_cfft_f64_cmplx_noisy_4096 there is a malloc() of length:
> 8192 * sizeof(float64_t) = 8192 * 8 = 64k
> but we need more RAM for variables, stack, ...

As you noted in #27849, `min_ram: 64` is indeed insufficient.

I am not sure if setting this to `128` is the best solution though; `80` or `96` should (at least, theoretically) work.

File: tests/lib/cmsis_dsp/transform/src/cf64.c [hunk 2]
```diff
@@ -167,42 +167,42 @@ DEFINE_TEST_VARIANT4(arm_cifft_f64_cmplx, step_4096, &arm_cfft_sR_f64_len4096,
...
+		ztest_unit_test(test_arm_cfft_f64_cmplx_step_256),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_step_256),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_step_128),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_step_128),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_step_64),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_step_64),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_step_32),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_step_32),
+		ztest_unit_test(test_arm_cfft_f64_cmplx_step_16),
+		ztest_unit_test(test_arm_cifft_f64_cmplx_step_16)
```
Comment (stephanosio) [inline tests/lib/cmsis_dsp/transform/src/cf64.c:205]: From #27849, re:
> I don't know exactly how malloc() algorithm from newlibc is working,
> but it appears during my debug session that even if memory is freed between each step,
> each time the size of the malloc is increased, the previous memory size is 'lost' (probably due to heap fragmentation).

IMO, this needs more investigation. I am not sure if fragmentation is the cause as far as this test goes (there aren't all that many mixed-size mallocs here to begin with -- all the allocated blocks are freed before subsequent different-size mallocs); rather, it seems there is something strange going on.

**Review by stephanosio — COMMENTED** [review]
Could you split the changes in `testcase.yml` and `cf64.c` into separate commits?

Ideally, updating the memory requirement in `testcase.yml` should be sufficient; but, it seems an additional change (reordering) is required due to something funny going on inside newlib malloc -- which ought to be investigated and fixed; after which, it should no longer be necessary and therefore reverted.

File: tests/lib/cmsis_dsp/transform/testcase.yaml (hunk 1 above)
Comment (ABOSTM) [reply tests/lib/cmsis_dsp/transform/testcase.yaml:50]: I dig a bit a found a nucleo_l476rg (96k ram 1024 flash) able to pass the test.
And I found no board (any manufacturers) with 80k ram and enough flash (1024)
So I propose to set min_ram: 96

Comment (ABOSTM) [reply tests/lib/cmsis_dsp/transform/testcase.yaml]: Split done
set 96 done

File: tests/lib/cmsis_dsp/transform/src/cf64.c (hunk 2 above)
Comment (ABOSTM) [reply tests/lib/cmsis_dsp/transform/src/cf64.c:205]: IMO fragmentation is the issue:
When test was in increasing order, the return of malloc() was respectively:
0x200017A8
0x200018B0
0x20001AB8
0x20001EC0
0x200026C8
0x200036D0 (step 512, malloc 0x2000)
0x200056D8 (step 1024, malloc 0x4000)
0x200096E0  (step 2048, malloc 0x8000)
Each time the malloc pointer is a bit more than the previous malloc.
I can just assume maaloc algo can reuse freed chunck for smaller malloc but cannot reuse for lager one.

**Comment by ABOSTM** [comment]
Similar PR for tests/lib/cmsis_dsp/matrix/: #27925

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@erwango @ioannisg please review

**Review by erwango — APPROVED** [review]

File: tests/lib/cmsis_dsp/transform/src/cf64.c (hunk 2 above)
Comment (ioannisg) [reply tests/lib/cmsis_dsp/transform/src/cf64.c:205]: @stephanosio ^^

**Review by ioannisg — COMMENTED** [review]
Even if the fragmentation is not the root cause if this problem (which this PR is trying to fix), I think this PR is fine as it is, theoretically, a "no-op" : it only re-orders the tests.

File: tests/lib/cmsis_dsp/transform/testcase.yaml [hunk 3]
```diff
@@ -47,7 +47,7 @@ tests:
       - CONFIG_CMSIS_DSP_TEST_TRANSFORM_RF32=y
   libraries.cmsis_dsp.transform.cf64:
     min_flash: 1024
-    min_ram: 64
+    min_ram: 96
```
Comment (ioannisg) [inline tests/lib/cmsis_dsp/transform/testcase.yaml:50]: @ABOSTM  Does this really need more than 64K of RAM in several ones or only this particular platform?
Could you try to build in some more ARM Boards with 64K RAM and see what is going on?

**Comment by ioannisg** [comment]
Not urgent to be merged before the 2.4 freeze - this is probably fixing a bug, so it can be dealt with later.

File: tests/lib/cmsis_dsp/transform/testcase.yaml (hunk 3 above)
Comment (ABOSTM) [reply tests/lib/cmsis_dsp/transform/testcase.yaml:50]: @ioannisg , as explained in https://github.com/zephyrproject-rtos/zephyr/issues/27849
there is a malloc of 64K so when you add stack, variables ... 
So for sure, 64K is not enough.

Comment (ioannisg) [reply tests/lib/cmsis_dsp/transform/testcase.yaml:50]: agreed, thanks

File: tests/lib/cmsis_dsp/transform/src/cf64.c (hunk 2 above)
Comment (stephanosio) [reply tests/lib/cmsis_dsp/transform/src/cf64.c:205]: Fair enough.

**Review by stephanosio — APPROVED** [review]

**Pull request merged**